src/logging.rs

```diff
@@ -103,11 +118,17 @@ fn is_valid_field(input: &str) -> bool {
 ///
 /// See <https://systemd.io/JOURNAL_NATIVE_PROTOCOL/> for details.
 fn add_field_and_payload_explicit_length(data: &mut Vec<u8>, field: &str, payload: &str) {
-    data.extend(field.as_bytes());
-    data.push(b'\n');
-    data.extend(&(payload.len() as u64).to_le_bytes());
-    data.extend(payload.as_bytes());
-    data.push(b'\n');
+    data.extend(
+        [
+            field.as_bytes(),
+            b"\n",
+            &(payload.len() as u64).to_le_bytes(),
+            payload.as_bytes(),
+            b"\n",
+        ]
+        .iter()
+        .flat_map(|x| x.iter()),
```

Comment:
Does this actually make much of a difference, allocation wise?  I'm sorry if this is a stupid question, but I'm not familiar with the allocation behaviour of vec, and the documentation seems to suggest that extend essentially just loops over `push`.  In this case the effect would be the same, but with–in my opinion–more obscure code.

If allocations are an issue, shouldn't we rather try to explicitly extend `data`s `capacity` to account for all these bytes?  This would also make the intend clearer.

And then, shouldn't we perhaps try to use a reasonably large capacity upfront, so that `data` ideally never needs to reallocate at all?

Reply:
Okay, I just saw the explanation in your PR description; sound reasonable, but… I find this a little implicit, to be honest.    If the express purpose is to ensure the vector has appropriate capacity I think we should write this down explicitly, i.e. keep the current code, and instead add an explicit `.reserve` call before.  After all, if `extend` can figure out the appropriate size, we can too.

If you and @lucab prefer the extend call, I'd really appreciate if it could get a small comment which explains that the code was written this way to reduce the number of allocations, and how it does achieve this goal.

Reply:
Yes, those are my feelings too.
Even without concrete number, I do trust @dfreese intuition that this is likely an hotpath where it may be worth reducing spurious allocations. At the same time, I do share @swsnr's position that the new code is harder to understand and it is not totally obvious that we are doing that dance for performance reasons.

Let's keep this change for a separate PR for now. It sounds like we are not in a hurry to change this, and I'd like to find some time to better evaluate this.

Reply:
I'm curious: Shouldn't an explicit reserve call before the current code gave the same effect?🤔  That'd be easy to read, and make the intention quite obvious 🙂

Reply:
reserve does work here.  You're right.  I was over thinking this.  I've gone ahead and added that since it seems like a clear win.